### ISSUE TITLE
[feature] MGH/MGZ support

### ISSUE THREAD

**Issue by NicolasRannou** [FNNDSC/ami#134]
https://github.com/aces/brainbrowser/tree/master/src/brainbrowser/volume-viewer/volume-loaders

**Comment by bwagnerwork** [comment]
I've got the initial version of a MGH/MGZ reader at https://github.com/bwagnerwork/ami/tree/freesurfer_reader . The two items I'm not sure of are the     imageOrientation and imagePosition.  I'd appreciate any review/comments.  Next is work on the surface reader (issue  #242 ) and when all is good submit a PR for both.

**Comment by NicolasRannou** [comment]
That was fast thanks!

The only comment so far is that you can remove the function that remove default values, for instance https://github.com/bwagnerwork/ami/blob/freesurfer_reader/src/parsers/parsers.mgh.js#L234-L236. Those functions returning the default values should already be defined at in parsers.volume.  https://github.com/FNNDSC/ami/blob/dev/src/parsers/parsers.volume.js

Would you  have some anonymized data we could use for an example?

Thanks!

**Comment by NicolasRannou** [comment]
For the work on the surface reader, we do not have any presets API yet and we write it the same way THREEJS does.

**Comment by bwagnerwork** [comment]
I've got the basic surface (triangle) reader coded.  Not quite ready for prime time as you can see.  I'll see about the quad after I get this cleaner.  Expect fixing this will take care of the imageOrientation and imagePosition I mentioned before.

![image](https://user-images.githubusercontent.com/9351579/35514274-45384eca-04cb-11e8-9a62-ab037d5cfdea.png)

Running FreeSurfer on their sample subjects.  I'll place that somewhere when done later today. I'll also clean the code as requested.

**Comment by NicolasRannou** [comment]
Awesome - about the model, could it also be that is is been displayed in RAS and not LPS space?

**Comment by bwagnerwork** [comment]
I've got the surface and MGH reader talking nicely together.  The MGH files I have are in LIA space.  I had to read the data in differently to get the overlay to sit right.  There is this note on the brainbrowser MGH reader:

    /** This is here because there are two different ways of interpreting
      * the origin of an MGH file. One can ignore the offsets in the
      * transform, using the centre of the voxel grid. Or you can correct
      * these naive grid centres using the values stored in the transform.
      * The first approach is what is used by surface files, so to get them
      * to register nicely, we want ignore_offsets to be true. However,
      * getting volumetric files to register correctly implies setting
      * ignore_offsets to false.
      */

![image](https://user-images.githubusercontent.com/9351579/35536012-28dddb4c-050b-11e8-9ebb-604c45c607d3.png)

I'm not entirely happy with the way I've done this.  The orientation of the images is wrong, but matches up with what is read in from the surface file.  I've modified the reading of the volume data to account for the surface.  Maybe it would be better to modify the reading of the surface and keep the volume unchanged.  Your thoughts?  I guess it is also a possibility to apply a matrix to the surface (as is done in loader_nifti), but it seems to me that if someone is loading FreeSurfer data they shouldn't have to do all that.

Changes are in my fork.

**Comment by bwagnerwork** [comment]
Just for kicks, the change to get it to display as I think it should is 3 different lines:

```
diff --git a/src/loaders/loaders.freesurfer.js b/src/loaders/loaders.freesurfer.js
index 22f1367..4059c45 100644
--- a/src/loaders/loaders.freesurfer.js
+++ b/src/loaders/loaders.freesurfer.js
@@ -122,7 +122,7 @@ Object.assign(THREE.FreeSurferLoader.prototype, THREE.EventDispatcher.prototype,
                 for (let v=0; v < vertCount; v++)
                 {
                     geometry.vertices.push(
-                        new THREE.Vector3(reader.getFloat32(offset +  0,littleEndian),  reader.getFloat32(offset +  4,littleEndian), reader.getFloat32(offset +  8,littleEndian) )
+                        new THREE.Vector3(reader.getFloat32(offset +  0,littleEndian) * -1,  reader.getFloat32(offset +  4,littleEndian) * -1, reader.getFloat32(offset +  8,littleEndian) )
                     );
                     offset+=12;
                 }
diff --git a/src/parsers/parsers.mgh.js b/src/parsers/parsers.mgh.js
index 970ecb5..b903019 100644
--- a/src/parsers/parsers.mgh.js
+++ b/src/parsers/parsers.mgh.js
@@ -211,8 +211,8 @@ export default class ParsersMgh extends ParsersVolume {
         
         if (this._readStyle==1)
         {
-            this._imageOrient=[1, 0, 0, 0, 0, -1];
-            this._origin=[-128, -128, 128];
+            this._imageOrient=[-1, 0, 0, 0, 0, -1];
+            this._origin=[128, 128, 128];
         } else {
             this._imageOrient = [
                 this._Xras[0], this._Xras[1], this._Xras[2],
```

**Comment by NicolasRannou** [comment]
Can you submit a PR so we can have a closer look over there?


That seems pretty encouraging though! Is the current contour visualization what you were looking for?

It seems the "meshes" from Freesurfer are in RAS space coordinates (like VTK does) all the time so after contccuting the mesh, it may just be a matter of applying a simple transform to flip axis:

From lesson 2:
```
var loaderSTL = new THREE.STLLoader();
loaderSTL.load('https://cdn.rawgit.com/FNNDSC/data/master/stl/adi_brain/WM.stl', function(geometry) {
    var material = new THREE.MeshPhongMaterial({ color: 0xf44336, specular: 0x111111, shininess: 200 });
    var mesh = new THREE.Mesh(geometry, material);
    // to LPS space
    var RASToLPS = new THREE.Matrix4();
    RASToLPS.set(-1, 0, 0, 0, 0, -1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1);
    mesh.applyMatrix(RASToLPS);
    scene.add(mesh);
});
```

**Comment by thewag1101** [comment]
Sample FreeSurfer running their sample subject at https://129.112.9.32/tmp/fs6_sample.tgz . #244 submitted.  I've not tried using the RAStoLPS example yet.

**Comment by NicolasRannou** [comment]
Closed by #244
